### PR TITLE
[SPARK-25458][SQL] Support FOR ALL COLUMNS in ANALYZE TABLE

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
**Description from the JIRA :** 
Currently, to collect the statistics of all the columns, users need to specify the names of all the columns when calling the command "ANALYZE TABLE ... FOR COLUMNS...". This is not user friendly. Instead, we can introduce the following SQL command to achieve it without specifying the column names.

```
   ANALYZE TABLE [db_name.]tablename COMPUTE STATISTICS FOR ALL COLUMNS;
```

## How was this patch tested?
Added new tests in SparkSqlParserSuite and StatisticsSuite

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 1]
```diff
@@ -33,11 +33,13 @@ import org.apache.spark.sql.types._
...
  * Analyzes the given columns of the given table to generate statistics, which will be used in
- * query optimizations.
+ * query optimizations. Parameter `allColumns` may be specified to generate statistics of all the
+ * columns of a given table.
  */
 case class AnalyzeColumnCommand(
     tableIdent: TableIdentifier,
-    columnNames: Seq[String]) extends RunnableCommand {
+    columnNames: Option[Seq[String]],
+    allColumns: Boolean = false ) extends RunnableCommand {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: nit. `false )` -> `false`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 2]
```diff
@@ -50,7 +52,26 @@ case class AnalyzeColumnCommand(
...
+      }
+    }
+    // Make sure the column types are supported for stats gathering.
+    attributesToAnalyze.foreach { attr =>
+      if (!supportsType(attr.dataType)) {
+        throw new AnalysisException(
+          s"Column ${attr.name} in table $tableIdent is of type ${attr.dataType}, " +
+            "and Spark does not support statistics collection on this column type.")
+      }
+    }
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: creating a new private function for the code between 55 and 72?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: Also throw an exception when `allColumns ` is set to true but columnNames is not empty.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 3]
```diff
@@ -70,25 +91,9 @@ case class AnalyzeColumnCommand(
    */
   private def computeColumnStats(
       sparkSession: SparkSession,
-      tableIdent: TableIdentifier,
-      columnNames: Seq[String]): (Long, Map[String, CatalogColumnStat]) = {
-
+      relation: LogicalPlan,
+      attributesToAnalyze: Seq[Attribute]): (Long, Map[String, CatalogColumnStat]) = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: `attributesToAnalyze ` -> `columns`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 4]
```diff
@@ -50,7 +52,26 @@ case class AnalyzeColumnCommand(
     val sizeInBytes = CommandUtils.calculateTotalSize(sparkSession, tableMeta)
 
     // Compute stats for each column
-    val (rowCount, newColStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+    val conf = sparkSession.sessionState.conf
+    val relation = sparkSession.table(tableIdent).logicalPlan
+    val attributesToAnalyze = if (allColumns) {
+      relation.output
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: Are we still able to create a table with zero column? for example, using dataframewriter?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala (hunk 1 above)
Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: let us do not use the default values?

**Comment by gatorsmile** [comment]
@dilipbiswal Thanks for working on this! Also cc @juliuszsompolski

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala (hunk 1 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: @gatorsmile ok.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: @gatorsmile OK

Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: @gatorsmile 
> creating a new private function for the code between 55 and 72
Actually, we may not be able to refactor from 55 to 72. We need to compute the `relation` once. I will do the following outside and call the proposed private method.
```
 val conf = sparkSession.sessionState.conf
 val relation = sparkSession.table(tableIdent).logicalPlan
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala (hunk 4 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: @gatorsmile I think we don't allow it any more. Please see [test](https://github.com/dilipbiswal/spark/blob/7920b29f667de38ca5222e4de8f62eea688d67e4/sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala#L111:L136). Please let me know what you think..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala [hunk 5]
```diff
@@ -48,9 +52,13 @@ case class AnalyzeColumnCommand(
       throw new AnalysisException("ANALYZE TABLE is not supported on views.")
     }
     val sizeInBytes = CommandUtils.calculateTotalSize(sparkSession, tableMeta)
+    val conf = sparkSession.sessionState.conf
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: useless?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 6]
```diff
@@ -653,6 +653,49 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+      val e1 = intercept[IllegalArgumentException] {
+        AnalyzeColumnCommand(TableIdentifier(table), Option(Seq("c1")), true).run(spark)
+      }
+      assert(e1.getMessage.contains("Parameter `columnNames` or `allColumns` are" +
+        " mutually exclusive"))
+      val e2 = intercept[IllegalArgumentException] {
+        AnalyzeColumnCommand(TableIdentifier(table), None, false).run(spark)
+      }
+      assert(e1.getMessage.contains("Parameter `columnNames` or `allColumns` are" +
+        " mutually exclusive"))
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Could we create a separate test case between 686 and 695

**Comment by gatorsmile** [comment]
LGTM except a few minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala (hunk 5 above)
Comment (dilipbiswal) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala]: @gatorsmile Yeah... Thanks.. will remove.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: @gatorsmile Sure.

**Comment by juliuszsompolski** [comment]
jenkins retest this

**Comment by juliuszsompolski** [comment]
LGTM. Thanks!

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**